Add note about pico-sdk-tools to README.md

will-v-pi opened the pull request:
Add note about pre-built picotool binaries in the pico-sdk-tools repository

will-v-pi committed:
Add note about pico-sdk-tools to README.md

lurch commented on README.md:
```diff
@@ -1,5 +1,7 @@
 ## Building
 
+> If you don't need to build picotool yourself, you can find pre-built binaries for Windows, macOS, and Linux in the [pico-sdk-tools](https://github.com/raspberrypi/pico-sdk-tools/releases) repository
```
"If you don't need..." sounds a bit strange - why would anybody **need** to build picotool themselves? :wink: 

Perhaps this could be shortened to something like "Pre-built binaries of picotool for Windows, macOS, and Linux can be downloaded from the [pico-sdk-tools](https://github.com/raspberrypi/pico-sdk-tools/releases) repository. But if you want to build it yourself from source..."

Hmm, it's obviously a slightly larger change, but given that we _are_ now offering pre-built binaries, perhaps the "Building" section should come _after_ the "Overview" section? Or perhaps be moved to an entirely separate `BUILDING.md` file?

will-v-pi replied:
I think that keeping the Building and Usage by the SDK sections at the top is useful, as the Overview is very long and therefore people might miss the build instructions, which would probably create more issues.

> "If you don't need..." sounds a bit strange - why would anybody **need** to build picotool themselves? 😉

You might not trust binaries you've downloaded from the internet, or you might want to install picotool system-wide - both of these would require building it yourself, although maybe "If you don't want..." would be better phrasing? I think it's still the main use case for people who arrive at this repository, and building it from source should work fine across all platforms - it works fine on Windows for me

lurch replied:
Fair enough. But perhaps it's worth having a brief "Introduction" or "Description" at the top of `README.md`, rather than diving straight into "Building" on the very first line? :shrug:

kilograham replied:
I agree with @lurch - possibly move the building into a separate BUILDING.md and link from the other

Note also I think the linked getting-started-doc has less info than is in the README,

Open to suggestion about where to put the "Use by SDK" stuff - probably in the BUILDING.md but note in the main README link that the BUILDING.md (should it be INSTALL.md) also explains how to build/install to prevent it needing to be installed on the fly - perhaps even just keep the section on use by the SDK, but just have it explain that SDK uses it, and you can install locally to not need it build every time, and then link to the BUILDING.md

will-v-pi replied:
I've restructured it, so the main README just has a small section linking to BUILDING.md, and includes the details about pico-sdk-tools. I've also added some hopefully helpful notes to prevent people doing the wrong thing (eg installing it incorrectly by just copying it).

will-v-pi committed:
Separate out BUILDING.md

Move build & install instructions into separate file, liked from the main readme

will-v-pi committed:
Remove references to `picotool sign`

Fixes #218

lurch commented on BUILDING.md:
````diff
@@ -0,0 +1,99 @@
+sudo apt install build-essential pkg-config libusb-1.0-0-dev cmake
+```
+
+> If libusb-1.0-0-dev is not installed, picotool still builds, but it omits all options that deal with managing a pico via USB (load, save, erase, verify, reboot). Builds that do not include USB support can be recognized because these commands also do not appear in the help command. The build output message 'libUSB is not found - no USB support will be built' also appears in the build logs.
````
Should this be moved out of the "Linux / macOS" section, and into the previous section where it  says "You also need to install `libusb-1.0` if you want to use the USB functionality." ?

lurch commented on BUILDING.md:
````diff
@@ -0,0 +1,99 @@
+Use your favorite package tool to install dependencies. For example, on Ubuntu:
+
+```console
+sudo apt install build-essential pkg-config libusb-1.0-0-dev cmake
````
Is there a simple `brew` command to install the needed dependencies on macOS?

will-v-pi replied:
Possibly just `brew install pkgconf libusb cmake`? But you also need to install the Xcode command line tools, or some alternative compiler that CMake can find

lurch replied:
Hmmm, in that case perhaps the "macOS" instructions should be broken out into their own section?

will-v-pi replied:
Maybe a separate PR to improve (and test) the MacOS instructions once this one has been merged in? This was just copied directly from the existing README

lurch commented on BUILDING.md:
```diff
@@ -0,0 +1,99 @@
+
+Download libUSB from here https://libusb.info/
+
+set LIBUSB_ROOT environment variable to the install directory.
```
nit: capital 'S' on "set" ? (also on line 54_

lurch commented on README.md:
```diff
@@ -155,6 +57,12 @@ Use "picotool help <cmd>" for more info
 
+## Building & Installing
+
+If you don't want to build picotool yourself, you can find pre-built executables for Windows, macOS, and Linux in the [pico-sdk-tools](https://github.com/raspberrypi/pico-sdk-tools/releases) repository. Assuming you've unzipped that directory to `$PICOTOOL_LOC` (with the actual picotool executable at `$PICOTOOL_LOC/picotool/picotool`), you can point the Pico SDK at this binary by setting the `pic ...
```
Maybe this would be slightly less confusing if you used `$SDKTOOLS_LOC` instead of `$PICOTOOL_LOC` ? (assuming of course that I haven't misunderstood what's going on! :wink: )

will-v-pi replied:
The releases contain `pico-sdk-tools` releases (currently just contains `pioasm`) and `picotool` releases as separate files, so I think `SDKTOOLS_LOC` might be confusing?

lurch replied:
Perhaps `$EXTRACT_LOCATION` ? My concern with `$PICOTOOL_LOC/picotool/picotool` is that people might find it confusing with the word "picotool" appearing 3 times?

lurch commented on README.md:
```diff
@@ -155,6 +57,12 @@ Use "picotool help <cmd>" for more info
 
+## Building & Installing
+
+If you don't want to build picotool yourself, you can find pre-built executables for Windows, macOS, and Linux in the [pico-sdk-tools](https://github.com/raspberrypi/pico-sdk-tools/releases) repository. Assuming you've unzipped that directory to `$PICOTOOL_LOC` (with the actual picotool executable at `$PICOTOOL_LOC/picotool/picotool`), you can point the Pico SDK at this binary by setting the `pic ...
```
"unzipped that directory" doesn't quite sound right?

will-v-pi replied:
Yeah, I wasn't really sure how to phrase it - I want to say that if you download it to Downloads/xxx.zip, extract that to Downloads/xxx, then the picotool binary is at Downloads/xxx/picotool/picotool - the crucial point is that it's not at Downloads/xxx/picotool

lurch replied:
I think that depending on the "unzip" program being used and/or which unzip options are chosen, you might end up with `picotool-2.1.1-x64-win\picotool\picotool.exe` or you might end up with just `picotool\picotool.exe` ?
Either way, I think "extracted that archive" might sound better than "unzipped that directory"?

will-v-pi committed:
Review fixups

Move no-libusb bit
Capital S
`$PICOTOOL_LOC` -> `$EXTRACT_LOCATION`
unzipped that directory -> extracted that archive

lurch commented on BUILDING.md:
```diff
@@ -8,6 +8,8 @@ git submodule update --init lib/mbedtls
 
 You also need to install `libusb-1.0` if you want to use the USB functionality.
 
+> If libusb-1.0 is not installed, picotool still builds, but it omits all options that deal with managing a pico via USB (load, save, erase, verify, reboot). Builds that do not include USB support can be recognized because these commands also do not appear in the help command. The build output message 'libUSB is not found - no USB support will be built' also appears in the build logs.
```
I think "because these commands also do not appear in the help command." should perhaps be reworded to "because these commands won't appear in the output of the help command." ?

will-v-pi committed:
Review fixup

because these commands also do not appear in the help command. -> because these commands won't appear in the output of the help command.

lurch commented on README.md:
```diff
@@ -155,6 +57,12 @@ Use "picotool help <cmd>" for more info
 
+## Building & Installing
+
+If you don't want to build picotool yourself, you can find pre-built executables for Windows, macOS, and Linux in the [pico-sdk-tools](https://github.com/raspberrypi/pico-sdk-tools/releases) repository. Assuming you've extracted that archive to `$EXTRACT_LOCATION` (with the actual picotool executable at `$EXTRACT_LOCATION/picotool/picotool`), you can point the Pico SDK at this binary by setting t ...
```
Reading this again, would using `<extract_location>` to signify a placeholder-directory work better than using `$EXTRACT_LOCATION`? Because otherwise some users might think that they need to set an `$EXTRACT_LOCATION` environment variable? :thinking:

will-v-pi replied:
Yeah, that's clearer
```suggestion
If you don't want to build picotool yourself, you can find pre-built executables for Windows, macOS, and Linux in the [pico-sdk-tools](https://github.com/raspberrypi/pico-sdk-tools/releases) repository. Assuming you've extracted that archive to `<extract_location>` (with the actual picotool executable at `<extract_location>/picotool/picotool`), you can point the Pico SDK at this binary by setting the `picotool_DIR` environment variable to `<extract_location>/picotool`, or by passing `-Dpicotool_DIR=<extract_location>/picotool` to your `cmake` command or setting it in your `CMakeLists.txt` file.
```

lurch reviewed (approved):
One small question, but otherwise LGTM

will-v-pi committed:
Update README.md

kilograham reviewed (approved):
nice

lurch commented on BUILDING.md:
````diff
@@ -0,0 +1,99 @@
+You need to set PICO_SDK_PATH in the environment, or pass it to cmake with `-DPICO_SDK_PATH=/path/to/pico-sdk`. To use features such as signing or hashing, you will need to make sure the mbedtls submodule in the SDK is checked out - this can be done by running this from your SDK directory.
+
+```console
+git submodule update --init lib/mbedtls
````
Hmmm, in light of other recent changes, does this now also need to update mbedtls's submodules too?

will-v-pi replied:
It won't once #226 is merged, so I'll leave this as is

lurch replied:
So now that **this** PR has been merged, does that imply that #226 ought to be rebased, and make its own changes to `README.md`? Or am I just getting confused by all the overlapping PRs again? :rofl:

will-v-pi replied:
#226 removes the need for mbedtls submodules, so won't need the readme update

lurch replied:
Right, so does #226 also need to update `BUILDING.md` to remove the `git submodule update` command, or am I still getting my wires crossed?

will-v-pi replied:
You still need the SDK's mbedtls submodule (pico-sdk/lib/mbedtls), but you don't need Mbedtls's submodules (pico-sdk/lib/mbedtls/framework etc) - so the command is still needed, but a new command to do it recursively instead is not needed, so no changes need to be made to the readme.

lurch replied:
Aha, thanks! It's the "you don't need Mbedtls's submodules (pico-sdk/lib/mbedtls/framework etc)" bit that I didn't realise, and is why I was getting confused :rofl:   But in that case, why is / was https://github.com/raspberrypi/pico-examples/pull/635 needed?

will-v-pi replied:
That works around the "problem" that #226 fixes, but probably isn't necessary if #226 will be merged soon, although won't do any harm either

will-v-pi replied:
Basically if you consider needing Mbedtls submodules to be the bug, then #226 is the bug-fix, and raspberrypi/pico-examples#635, https://github.com/raspberrypi/pico-sdk/pull/2416, 8be60e9eeab96857c84d71f0565cad0e84654c77 are all workarounds to get CI working
